### PR TITLE
Fix end-to-end test logic to get image name

### PR DESCRIPTION
**Pull request by maht**
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
It is needed to support multiple architectures when running the end-to-end tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
None, as far as I know.

**Special notes for your reviewer**:
This was tested for ARM64 for conformance tests, using Kubernetes v1.12.1 (Docker v18.03.1-ce, Ubuntu 18.04.4)

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

**Comment by maht** [comment]
It seems that the job failed due to a panic in `containerd` process, producing a transient failure in the tests (see log analysis below). Retesting to verify it:

/retest

[failed test log](https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/70648/pull-kubernetes-e2e-kops-aws/112375/#sig-storage-emptydir-volumes-when-fsgroup-is-specified-nodefeaturefsgroup-volume-on-tmpfs-should-have-the-correct-mode-using-fsgroup):

```
STEP: Collecting events from namespace "e2e-tests-emptydir-np87f".
STEP: Found 4 events.
Nov  6 01:11:12.160: INFO: At 2018-11-06 01:11:09 +0000 UTC - event for pod-d812f621-e160-11e8-ac26-0a580a3c0a2f: {default-scheduler } Scheduled: Successfully assigned e2e-tests-emptydir-np87f/pod-d812f621-e160-11e8-ac26-0a580a3c0a2f to ip-172-20-41-33.eu-west-1.compute.internal
Nov  6 01:11:12.160: INFO: At 2018-11-06 01:11:10 +0000 UTC - event for pod-d812f621-e160-11e8-ac26-0a580a3c0a2f: {kubelet ip-172-20-41-33.eu-west-1.compute.internal} Pulled: Container image "gcr.io/kubernetes-e2e-test-images/mounttest:1.0" already present on machine
Nov  6 01:11:12.160: INFO: At 2018-11-06 01:11:10 +0000 UTC - event for pod-d812f621-e160-11e8-ac26-0a580a3c0a2f: {kubelet ip-172-20-41-33.eu-west-1.compute.internal} Created: Created container
Nov  6 01:11:12.160: INFO: At 2018-11-06 01:11:10 +0000 UTC - event for pod-d812f621-e160-11e8-ac26-0a580a3c0a2f: {kubelet ip-172-20-41-33.eu-west-1.compute.internal} Failed: Error: failed to start container "test-container": Error response from daemon: grpc: the connection is unavailable

Nov  6 01:11:12.377: INFO: POD                                                                 NODE                                         PHASE    GRACE  CONDITIONS
...
Nov  6 01:11:12.378: INFO: pod-subpath-test-aws-dynamicpv-gjrp                                 ip-172-20-37-131.eu-west-1.compute.internal  Pending         [{Initialized False 0001-01-01 00:00:00 +0000 UTC 2018-11-06 01:11:01 +0000 UTC ContainersNotInitialized containers with incomplete status: [init-volume-aws-dynamicpv-gjrp]} {Ready False 0001-01-01 00:00:00 +0000 UTC 2018-11-06 01:11:01 +0000 UTC ContainersNotReady containers with unready status: [test-container-subpath-aws-dynamicpv-gjrp test-container-volume-aws-dynamicpv-gjrp]} {ContainersReady False 0001-01-01 00:00:00 +0000 UTC 2018-11-06 01:11:01 +0000 UTC ContainersNotReady containers with unready status: [test-container-subpath-aws-dynamicpv-gjrp test-container-volume-aws-dynamicpv-gjrp]} {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2018-11-06 01:11:01 +0000 UTC  }]
```


[kubelet log at  ip-172-20-41-33.eu-west-1.compute.internal node](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/70648/pull-kubernetes-e2e-kops-aws/112375/artifacts/ip-172-20-41-33.eu-west-1.compute.internal/kubelet.log):

```
E1106 01:11:10.496375    1648 remote_runtime.go:213] StartContainer "388bf3ee747a3858bb6fa26b4cb16c09d476741cd9b8cd78ad48e0aa6b663d73" from runtime service failed: rpc error: code = Unknown desc = failed to start container "388bf3ee747a3858bb6fa26b4cb16c09d476741cd9b8cd78ad48e0aa6b663d73": Error response from daemon: grpc: the connection is unavailable
```


[docker log at  ip-172-20-41-33.eu-west-1.compute.internal](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/70648/pull-kubernetes-e2e-kops-aws/112375/artifacts/ip-172-20-41-33.eu-west-1.compute.internal/docker.log):

```
time="2018-11-06T01:11:08.922355097Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 1b47862715347d39fef5ec4c43a5f671234ac8701f18a0466b61670c85383726: rpc error: code = 2 desc = no such process"
time="2018-11-06T01:11:08.922447764Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 6f0a53d0b62f3a1ae1db6d16ef48cb143f00f0cea8ac8725931135eded45cc7d: rpc error: code = 2 desc = no such process"
time="2018-11-06T01:11:09.095555595Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 287003975b9cf81b9a9c9c053e93da2af66f4ca631b73766b2a1cd08492fa778: rpc error: code = 2 desc = containerd: container not found"
time="2018-11-06T01:11:10.304319176Z" level=error msg="containerd: get exit status" error="containerd: process has not exited" id=cb72a6ceeb28b001a29d809bd2bed21306e940aeb180f92916593d325ec2a55d pid=455a2f2b643d2b5e24de9e964bccea2f943a544868205926fae155bf65abf103 systemPid=2448
panic: close of nil channel
goroutine 4645 [running]:
panic(0x560e680883e0, 0xc42030fee0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/docker/containerd/supervisor.(*Supervisor).execExit.func1(0xc4202841e0, 0xc42012eea0, 0x0)
	/tmp/tmp.jSFRxEilBP/src/github.com/docker/containerd/supervisor/exit.go:90 +0x10c
created by github.com/docker/containerd/supervisor.(*Supervisor).execExit
	/tmp/tmp.jSFRxEilBP/src/github.com/docker/containerd/supervisor/exit.go:91 +0xee
time="2018-11-06T01:11:10.310620118Z" level=error msg="libcontainerd: failed to receive event from containerd: rpc error: code = 13 desc = transport is closing"
```

**Comment by maht** [comment]
Something different went wrong in the last test. It seem it failed during test cluster setup. Retrying:

/retest

[log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/70648/pull-kubernetes-e2e-kops-aws/112478/build-log.txt):
```
W1106 14:55:15.502] 2018/11/06 14:55:15 process.go:153: Running: kubectl get nodes -ojson
W1106 14:55:15.631] Unable to connect to the server: dial tcp: lookup api.e2e-112478-dba53.test-cncf-aws.k8s.io on 10.63.240.10:53: no such host
...
W1106 15:16:06.454] 2018/11/06 15:16:06 main.go:314: Something went wrong: starting e2e cluster: waiting for ready nodes timed out
```

File: test/e2e/framework/rc_util.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func ValidateController(c clientset.Interface, containerImage string, replicas i
 	// You can read about the syntax here: http://golang.org/pkg/text/template/.
 	getContainerStateTemplate := fmt.Sprintf(`--template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if (and (eq .name "%s") (exists . "state" "running"))}}true{{end}}{{end}}{{end}}`, containername)
 
-	getImageTemplate := fmt.Sprintf(`--template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if eq .name "%s"}}{{.image}}{{end}}{{end}}{{end}}`, containername)
+	getImageTemplate := fmt.Sprintf(`--template={{if (exists . "spec" "containers")}}{{range .spec.containers}}{{if eq .name "%s"}}{{.image}}{{end}}{{end}}{{end}}`, containername)
```
Comment (timothysc) [inline test/e2e/framework/rc_util.go]: This code is legacy, and it's a bit obtuse to follow why this change is needed for multi-arch.  I'm not following why it doesn't just work.

Comment (maht) [reply test/e2e/framework/rc_util.go]: No problem, I can give more context.

I run to this problem because one of our end-to-end test in ARM was failing, emitting this error:

```
Oct 26 11:05:31.784: INFO: update-demo-kitten-7gx94 is created but running wrong image; expected: gcr.io/kubernetes-e2e-test-images/kitten:1.0, actual: gcr.io/kubernetes-e2e-test-images/kitten-arm64:1.0
```

As you can see, the image expected does not include the '-arm64' arch suffix.

It seems that the test does no include it, since they are platform agnostic, so the image name is a generic one (the one without the '-arm64' suffix). But this image name is a multiarch one, so this name only points to a manifest, that include the real image name for all the archs that are actually supported. For example:

```
$ kubectl describe pod -n e2e-tests-kubectl-fdqr8 update-demo-kitten-zj424
Name:               update-demo-kitten-zj424
Namespace:          e2e-tests-kubectl-fdqr8
Priority:           0
PriorityClassName:  <none>
Node:               x-5/192.168.2.55
Start Time:         Fri, 26 Oct 2018 13:38:26 +0000
Labels:             name=update-demo
                   version=kitten
Annotations:        midonet.org/mac-address=42:88:73:f6:38:5e
Status:             Running
IP:                 10.1.1.136
Controlled By:      ReplicationController/update-demo-kitten
Containers:
 update-demo:
   Container ID:   docker://3890dad2629483535ff1e86e25dd4e36f33415626fc85c35ff331950969aad21
   Image:          gcr.io/kubernetes-e2e-test-images/kitten:1.0
   Image ID:       docker-pullable://gcr.io/kubernetes-e2e-test-images/kitten-arm64@sha256:fbc805315cd3fc3adfa90848959193ae45808444e1eb9af855d32f1dfe11ded6
```

You can see that the value of the 'Image' field is different than the 'Image ID', which includes the actual image name and its SHA256 hash.

The code in the test seems to be using not the image name in the specification (under '.spec') of the container, but the image in the status (under '.status'), as can be seen by this log of the command that retrieves it (in kubernetes v1.12.1):

```
Oct 26 13:43:01.540: INFO: Running '/usr/bin/kubectl --kubeconfig=/home/nvidia/jenkins/workspace/iessp/iaas/kubernetes-testing/midonet_kubernetes_e2e_arm64/46/iessp-iaas/tests/ci/k8s/clusters/sshlogin/kubeconfig get pods update-demo-kitten-hsgdw -o template --template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if eq .name "update-demo"}}{{.image}}{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-fdqr8'
Oct 26 13:43:01.931: INFO: stderr: ""
Oct 26 13:43:01.931: INFO: stdout: "gcr.io/kubernetes-e2e-test-images/kitten-arm64:1.0"
```

Since the name retrieved by the command includes the '-arm64' suffix, the test code seems it is wrong.

I don't know if nobody is using this code (kubernetes v1.12.1) on ARM64, or if this happening just in specific versions of the container runtime (specific Docker versions).

Is this code legacy now? no longer used in master? I didn't know, sorry. With this changes our tests pass correctly.

Anyway, I think the test code is more platform agnostic if the command retrieving the image name uses the 'spec' instead of the 'status', since the test code is no able to guess what is the name of the specific images per arch.

Thanks for your time. Let me know what you think and if anything else needs clarification.

Comment (timothysc) [reply test/e2e/framework/rc_util.go]: This makes perfect sense now. 
/lgtm 
/approve

**Review by timothysc — APPROVED** [review]
